### PR TITLE
Batch create users, and refactor of some types.

### PR DESCRIPTION
**Pull request by fisx**
commits should be reasonable self-explanatory.

### CODE REVIEW

**Comment by fisx** [comment]
Except for the fact that travis currently does not work, this is ready for review.

**Comment by np** [comment]
LGTM but my reading was shallow

File: src/Api/Persistent.hs [hunk 1]
```diff
@@ -241,11 +247,11 @@ findWildIdeasBySpace space = findAllIn dbIdeas (\idea -> idea ^. ideaSpace == sp
 
 -- | FIXME: anyone can login
 -- | FIXME: every login changes all other logins
-loginUser :: ST -> Persist ()
+loginUser :: UserLogin -> Persist ()
 loginUser login = modifyDb dbCurrentUser . const . fmap (view _Id) =<< findUserByLogin login
 
-logoutUser :: ST -> Persist ()
-logoutUser _login = modifyDb dbCurrentUser $ const Nothing
+logoutUser :: Persist ()
```
Comment (andorp) [inline src/Api/Persistent.hs]: I disagree here. The `currentUser` in the DB is a hack. We should not build our abstraction on it.

Comment (fisx) [inline src/Api/Persistent.hs]: Ok, get it.  Will fix.

File: src/Types.hs [hunk 2]
```diff
@@ -260,20 +260,40 @@ phaseName = \case
...
-type instance Proto User = User
+newtype UserLogin     = UserLogin     { _fromUserLogin     :: ST }
+  deriving (Eq, Ord, Show, Read, IsString, Monoid, Generic)
+
+newtype UserFirstName = UserFirstName { _fromUserFirstName :: ST }
+  deriving (Eq, Ord, Show, Read, IsString, Monoid, Generic)
+
+newtype UserLastName  = UserLastName  { _fromUserLastName  :: ST }
+  deriving (Eq, Ord, Show, Read, IsString, Monoid, Generic)
+
```
Comment (Mikolaj) [inline src/Types.hs]: Good idea with all these newtypes.

File: src/Types.hs [hunk 3]
```diff
@@ -328,10 +349,20 @@ newtype AUID a = AUID Integer
...
+--
+-- Some of these fields, like login name and avatar url of creator, are redundant.  The reason to
+-- keep them here is that it makes it easy to keep large 'Page*' types containing many nested
+-- objects, and still allowing all these objects to be rendered purely only based on the information
+-- they contain.
+--
+-- If this is becoming too much in the future and we want to keep objects around without all this
+-- inlined information, we should consider making objects polymorphic in the concrete meta info
+-- type.  Example: 'Idea MetaInfo', but also 'Idea ShortMetaInfo'.
 data MetaInfo a = MetaInfo
```
Comment (Mikolaj) [inline src/Types.hs]: Thanks for this haddock. Makes sense.

File: src/Types.hs [hunk 4]
```diff
@@ -328,10 +349,20 @@ newtype AUID a = AUID Integer
...
+-- | General information on objects stored in the DB.
+--
+-- Some of these fields, like login name and avatar url of creator, are redundant.  The reason to
+-- keep them here is that it makes it easy to keep large 'Page*' types containing many nested
+-- objects, and still allowing all these objects to be rendered purely only based on the information
+-- they contain.
+--
+-- If this is becoming too much in the future and we want to keep objects around without all this
+-- inlined information, we should consider making objects polymorphic in the concrete meta info
+-- type.  Example: 'Idea MetaInfo', but also 'Idea ShortMetaInfo'.
```
Comment (Mikolaj) [inline src/Types.hs]: Thanks for this haddock. Makes sense.

File: src/Frontend/Page/FileUpload.hs [hunk 5]
```diff
@@ -58,7 +59,43 @@ instance FormPageView BatchCreateUsers where
 theOnlySchoolYearHack :: Int
 theOnlySchoolYearHack = 2016
 
-type CsvUserRecord = [SBS]
+data CsvUserRecord = CsvUserRecord
+    { _csvUserRecordFirst       :: ST
+    , _csvUserRecordLast        :: ST
+    , _csvUserRecordEmail       :: Maybe ST
+    , _csvUserRecordLogin       :: Maybe ST
+    }
```
Comment (Mikolaj) [inline src/Frontend/Page/FileUpload.hs]: Probably a premature refactoring, but ideally we'd use the newtypes here, as well. And similarly in UserSettingData.

**Comment by Mikolaj** [comment]
LGTM.

**Pull request merged**